File: backend/app/pipeline/unzip.py
"""Pipeline step 1 — recursive unzip + dedup.

A DCE arrives as a ZIP (often with nested ZIPs) or as loose files. We walk it
recursively, keep only the document formats we can parse (PDF/DOCX/XLSX), and
drop byte-identical duplicates (the same annexe shipped in several sub-folders
is common).
"""

from __future__ import annotations

import hashlib
import io
import logging
import posixpath
import zipfile
from dataclasses import dataclass

log = logging.getLogger(__name__)

SUPPORTED_SUFFIXES = (".pdf", ".docx", ".xlsx")
_MAX_ZIP_DEPTH = 8


@dataclass
class ExtractedFile:
    """A single parseable document pulled out of the upload."""

    filename: str  # display name (archive-relative path, or the upload name)
    content: bytes
    sha256: str


def _is_supported(name: str) -> bool:
    return name.lower().endswith(SUPPORTED_SUFFIXES)


def _is_ignored(name: str) -> bool:
    base = posixpath.basename(name)
    # macOS resource forks / hidden files.
    return "__MACOSX" in name or base.startswith("._") or base.startswith(".")


def extract_upload(data: bytes, original_filename: str) -> list[ExtractedFile]:
    """Return the deduplicated list of parseable documents in ``data``.

    ``data`` is either a ZIP or a single loose document. Nested ZIPs are
    expanded recursively.
    """
    seen: set[str] = set()
    out: list[ExtractedFile] = []

    def add(name: str, content: bytes) -> None:
        digest = hashlib.sha256(content).hexdigest()
        if digest in seen:
            log.info("skip duplicate: %s", name)
            return
        seen.add(digest)
        out.append(ExtractedFile(filename=name, content=content, sha256=digest))

    def walk_zip(blob: bytes, prefix: str, depth: int) -> None:
        if depth > _MAX_ZIP_DEPTH:
            log.warning("max zip depth reached under %s", prefix)
            return
        try:
            zf = zipfile.ZipFile(io.BytesIO(blob))
        except zipfile.BadZipFile:
            log.warning("not a valid zip: %s", prefix)
            return
        for info in zf.infolist():
            if info.is_dir() or _is_ignored(info.filename):
                continue
            name = posixpath.join(prefix, info.filename) if prefix else info.filename
            try:
                content = zf.read(info)
            except Exception:  # noqa: BLE001 — a bad entry shouldn't sink the rest
                log.exception("failed to read zip entry %s", name)
                continue
            lower = info.filename.lower()
            if lower.endswith(".zip"):
                walk_zip(content, name, depth + 1)
            elif _is_supported(lower):
                add(name, content)

    if original_filename.lower().endswith(".zip") or zipfile.is_zipfile(io.BytesIO(data)):
        walk_zip(data, prefix="", depth=0)
    elif _is_supported(original_filename):
        add(original_filename, data)
    else:
        log.warning("unsupported top-level upload: %s", original_filename)

    return out

File: site-presentation/src/lib/dce/unzip.ts
// Pipeline step 1 — recursive unzip + dedup (server-only).
//
// A DCE arrives as a ZIP (often with nested ZIPs) or as loose files. We walk it
// recursively, keep only the formats we can parse (PDF/DOCX/XLSX + image files
// handled via OCR), and drop byte-identical duplicates (the same annexe shipped
// in several sub-folders is common).

import crypto from "node:crypto";

import { unzipSync } from "fflate";

import { IMAGE_EXTS } from "./extract";

const SUPPORTED = [".pdf", ".docx", ".xlsx", ...IMAGE_EXTS];
const MAX_ZIP_DEPTH = 8;

export type ExtractedFile = {
  filename: string; // archive-relative path, or the upload name
  content: Buffer;
  sha256: string;
};

function isSupported(name: string): boolean {
  const lower = name.toLowerCase();
  return SUPPORTED.some((s) => lower.endsWith(s));
}

function isIgnored(name: string): boolean {
  const base = name.split("/").pop() ?? name;
  return name.includes("__MACOSX") || base.startsWith("._") || base.startsWith(".");
}

function looksLikeZip(buf: Buffer): boolean {
  // Local file header "PK\x03\x04" or empty-archive "PK\x05\x06".
  return buf.length >= 4 && buf[0] === 0x50 && buf[1] === 0x4b && (buf[2] === 3 || buf[2] === 5);
}

export function extractUpload(data: Buffer, originalFilename: string): ExtractedFile[] {
  const seen = new Set<string>();
  const out: ExtractedFile[] = [];

  function add(name: string, content: Buffer) {
    const digest = crypto.createHash("sha256").update(content).digest("hex");
    if (seen.has(digest)) return;
    seen.add(digest);
    out.push({ filename: name, content, sha256: digest });
  }

  function walkZip(blob: Buffer, prefix: string, depth: number) {
    if (depth > MAX_ZIP_DEPTH) return;
    let entries: Record<string, Uint8Array>;
    try {
      entries = unzipSync(new Uint8Array(blob));
    } catch {
      return; // not a valid zip; skip
    }
    for (const [entryName, bytes] of Object.entries(entries)) {
      if (!bytes || bytes.length === 0) continue; // directory entry
      if (isIgnored(entryName)) continue;
      const name = prefix ? `${prefix}/${entryName}` : entryName;
      const content = Buffer.from(bytes);
      const lower = entryName.toLowerCase();
      if (lower.endsWith(".zip")) {
        walkZip(content, name, depth + 1);
      } else if (isSupported(lower)) {
        add(name, content);
      }
    }
  }

  if (originalFilename.toLowerCase().endsWith(".zip") || looksLikeZip(data)) {
    walkZip(data, "", 0);
  } else if (isSupported(originalFilename)) {
    add(originalFilename, data);
  }

  return out;
}
